devicetree: make DT_..._REG_ADDR return unsigned

fabiobaltieri opened the pull request:
`DT_..._REG_ADDR` macros do not always return an unsigned value. This commit adds and unsigned 32bit prefix to ensure the value is always unsigned.

More context:
- https://github.com/zephyrproject-rtos/zephyr/issues/53505
- https://github.com/zephyrproject-rtos/zephyr/pull/78419

Picking up from https://github.com/zephyrproject-rtos/zephyr/pull/70761

fabiobaltieri committed:
devicetree: make DT_..._REG_ADDR return unsigned

`DT_..._REG_ADDR` macros do not always return an unsigned value. This
commit adds and unsigned 32bit prefix to ensure the value is always
unsigned.

Signed-off-by: Jeppe Odgaard <jeppe.odgaard@prevas.dk>
Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

swift-tk commented:
The register address value are always unsigned, so maybe just use the new one with the cast instead of keeping the old one.
I believe this also applies to lot of other fields like "size".

jeppenodgaard commented on include/zephyr/devicetree.h:
```diff
@@ -2243,14 +2243,28 @@
+#define DT_REG_ADDR_BY_IDX_INTERNAL(node_id, idx) \
+	DT_CAT4(node_id, _REG_IDX_, idx, _VAL_ADDRESS)
+
+#define DT_REG_ADDR_INTERNAL(node_id) \
```
I think usage information should part be doxygen comment, so it will be used correctly.

Would be nice if the name could reflect that it does not prepend a prefix and/or that it is used to get generated define AFAIK.

fabiobaltieri replied:
yeah naming for this one could use more ideas, this should go for arch wg anyway imo, let me tag it so it gets blocked as well

henrikbrixandersen replied:
I'd like to see some documentation as well. How come this is internal but still used outside this file, for instance?

jeppenodgaard reviewed (commented):
Really nice to get this fixed. Well done 👍

fabiobaltieri commented:
> The register address value are always unsigned, so maybe just use the one with the cast instead of using the new one.

Not sure I'm following, what are you referring to? The one I've replaced around the code based are used to refer to macro themselves so they are not happy with the prefix, need to write a paragraph of description about it.

> I believe this also applies to lot of other fields like "size".

Yeah thinking about it though it's less likely to cause problems for size, it may be a good followup. Was mainly interested to see if I can get CI to pass here and open the conversation.

henrikbrixandersen reviewed (changes requested):
Please add some documentation for the introduced macros.

swift-tk commented:
> > The register address value are always unsigned, so maybe just use the one with the cast instead of using the new one.
> 
> Not sure I'm following, what are you referring to? The one I've replaced around the code based are used to refer to macro themselves so they are not happy with the prefix, need to write a paragraph of description about it.

The new one is actually renamed from the old one, naturally you would swap the ones in the headers. What I meant was that since register address are always unsigned, it was expected that DT_REG_ADDR* should give an unsigned value. So the headers should be using the one with unsigned cast.

My other thought is why don't we fix this in the generated macro script by adding the unsigned postfix "U" instead? given that lots of other unsigned values from devicetree are actually implied.

marc-hb commented on drivers/flash/flash_simulator.c:
```diff
@@ -174,9 +174,9 @@ static int flash_range_is_valid(const struct device *dev, off_t offset,
 {
 	ARG_UNUSED(dev);
-	if ((offset + len > FLASH_SIMULATOR_FLASH_SIZE +
+	if (((int64_t)(offset + len) > FLASH_SIMULATOR_FLASH_SIZE +
```
```suggestion
	if ((int64_t)offset + (int64_t)len >
```
A bit more verbose but avoid integer promotion rules madness entirely.

fabiobaltieri replied:
thanks, applied

gmarull commented on drivers/w1/w1_ds2482-800_channel.c:
```diff
@@ -164,8 +164,8 @@ static const struct w1_driver_api ds2482_driver_api = {
 		.i2c_spec = I2C_DT_SPEC_GET(DT_INST_PARENT(inst)),                                 \
-		.reg_channel = UTIL_CAT(CHSL_IO, DT_INST_REG_ADDR(inst)),                          \
-		.reg_channel_rb = UTIL_CAT(CHSL_RB_IO, DT_INST_REG_ADDR(inst)),                    \
+		.reg_channel = UTIL_CAT(CHSL_IO, DT_REG_ADDR_INTERNAL(DT_DRV_INST(inst))),         \
+		.reg_channel_rb = UTIL_CAT(CHSL_RB_IO, DT_REG_ADDR_INTERNAL(DT_DRV_INST(inst))),   \
```
this feels wrong:

1. using `_INTERNAL` externally doesn't look great
2. Using DT_DRV_INST should be always avoided: add missing _INST macros instead

fabiobaltieri replied:
Yeah the `INTERNAL` thing was more about hopes and dreams of that actually being internal, we know how that ended, changed to `RAW`, open to better suggestions, I suck bad with naming.

Added ther INST variants, current state was more like "can this even pass CI?".

fabiobaltieri commented:
> The new one is actually renamed from the old one, naturally you would swap the ones in the headers. What I meant was that since register address are always unsigned, it was expected that DT_REG_ADDR* should have gave an unsigned value. So the headers should be using the one with unsigned cast.

Sorry still confused, are you referring to the mspi change where I used the _INTERNAL one? That does not really work out, that macro wants just a number and ends up string matching it with the index of the gpio in the chip select gpios phandle-array. The explicit addresses (with the 'U' suffix) only work in C expression but unfortunately some macro use them for offset where it's effectively a string so I can't just use the suffixed version everywhere.

> My other thought is : why don't we fix this in the generated macro script by adding the unsigned postfix "U" instead? given that lots of other unsigned values from devicetree are actually implied. What is point of having an DT_REG_ADDR_INTERNAL that potentially give a signed value and that everyone should avoid using?

Because you may want to use the address as an index in a macro expansion that is really not going to like that suffix.

marc-hb commented:
> Because you may want to use the address as an index in a macro expansion that is really not going to like that suffix.

I don't whether it applies here, but FWIW assembly does not like `U` suffixes either

- https://github.com/thesofproject/sof/pull/9413#issuecomment-2318247608

fabiobaltieri commented:
> I don't whether it applies here, but FWIW assembly does not like `U` suffixes either

Yeah but the macro already has a no-op wrapper `#if defined(_LINKER) || defined(_ASMLANGUAGE)`, checked out a build for `mr_canhubk3` which has a `DT_REG_ADDR` in `soc/nxp/s32/s32k3/s32k3xx_startup.S` and it built happily, let me know if I'm missing anything but I think this is sorted already.

fabiobaltieri commented:
Okay main changes:

- renamed the `INTERNAL` macro to `RAW`, moved them off the internal doc block and added some doc
- added the missing `INST` varants

fabiobaltieri commented:
- unmessed the rebase (the flash change was in the wrong commit)

fabiobaltieri commented:
...and yes if we are happy with this it makes sense to do it for the SIZE macro as well, will do as a followup, just trying to see if this actually gets through

fgrandel commented:
Naive question: Could the casts not be done outside the macro where required? This would be more explicit and we won't get lots of additional macros that only differ by a cast - such a pattern doesn't scale very well, right? I mean people have to know what they're doing anyway - otherwise they could choose the wrong macro, too.

Side note: You're fixing offset/addr of the flash simulator. That seems to be a bug anyway as offsets are generated relatively to the base by the callers. This code only works because the base address is set to zero. All references to the base address should be removed there in favor of relative addressing (as the field provides anyway). (Checked this with the NVS settings subsys / NVS recently.)

fabiobaltieri commented:
> Naive question: Could the casts not be done outside the macro where required? This would be more explicit and we won't get lots of additional macros that only differ by a cast - such a pattern doesn't scale very well, right? I mean people have to know what they're doing anyway - otherwise they could choose the wrong macro, too.

Yes it could, not it shouldn't, changing the 1400+ instanced of this in the code and asking the reviewers to keep this in mind in addition to the other million things that they already have to keep in mind does not scale well, having the default macro doing the right thing and extra ones for when it does not work and the compiler tell you that it does not work scales well.

Yes people have to know what they are doing, yes they will still chose the wrong one, that's not a reason to not have defensivneess coded in the lower layers.

de-nordic reviewed (dismissed):
Flash Simulator looks ok.

fgrandel commented:
> Flash Simulator looks ok.

Don't think so, see https://github.com/zephyrproject-rtos/zephyr/issues/79082

carlescufi commented:
Arch WG:

- Overview of the change, no objections
- @decsny any objections to this being merged?

decsny commented on doc/releases/migration-guide-4.0.rst:
```diff
@@ -35,6 +35,13 @@ Boards
+**********
+
+* The :c:macro:`DT_REG_ADDR` macro and its variants are now expanding into an
+  unsigned literals (i.e. with w ``U`` suffix). To use addresses as devicetree
```
is this a typo? "with w" ?

fabiobaltieri replied:
it sure is, fixed

fabiobaltieri commented:
- fixed a typo in release notes

decsny commented on include/zephyr/devicetree.h:
```diff
@@ -2243,14 +2243,42 @@
+ * be used to index other devicetree properties, use the non _RAW macros for
+ * assigning values in actual code.
+ *
+ * Equivalent to DT_REG_ADDR_BY_IDX(node_id, 0).
```
another typo? DT_REG_ADDR_BY_IDX_RAW?

fabiobaltieri replied:
none of the second checking, all of the copypaste

aescolar commented:
@fabiobaltieri rebase needed

fabiobaltieri commented:
Merge conflict, rebased, I dropped the flash_simulator commit entirely as the validation code has been changed and I **think** the casting is not needed anymore, hang on for the CI run (can't recall what platform/combo test was failing specifically) will reapply if it's still a problem.

bjarki-andreasen reviewed (approved):
CI seems happy :)

fabiobaltieri commented:
hurray!
